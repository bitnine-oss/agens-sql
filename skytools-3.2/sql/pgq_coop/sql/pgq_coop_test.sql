
select pgq.create_queue('testqueue');
update pgq.queue set queue_ticker_max_count = 1 where queue_name = 'testqueue';

-- register
select pgq_coop.register_subconsumer('testqueue', 'maincons', 'subcons1');
select pgq_coop.register_subconsumer('testqueue', 'maincons', 'subcons1');
select pgq_coop.register_subconsumer('testqueue', 'maincons', 'subcons1');
select pgq_coop.register_subconsumer('testqueue', 'maincons', 'subcons2');

-- process events
select pgq_coop.next_batch('testqueue', 'maincons', 'subcons1');
select pgq.insert_event('testqueue', 'ev0', 'data');
select pgq.insert_event('testqueue', 'ev1', 'data');
select pgq.insert_event('testqueue', 'ev2', 'data');
select pgq.ticker();

select pgq_coop.next_batch('testqueue', 'maincons', 'subcons1');
select pgq_coop.next_batch('testqueue', 'maincons', 'subcons1');

select pgq_coop.next_batch('testqueue', 'maincons', 'subcons2');

select pgq.insert_event('testqueue', 'ev3', 'data');
select pgq.insert_event('testqueue', 'ev4', 'data');
select pgq.insert_event('testqueue', 'ev5', 'data');
select pgq.ticker();
select pgq_coop.next_batch('testqueue', 'maincons', 'subcons2');

select pgq_coop.finish_batch(2);

-- test takeover
select pgq_coop.next_batch('testqueue', 'maincons', 'subcons2', '1 hour');
update pgq.subscription set sub_active = '2005-01-01' where sub_batch is not null;
select pgq_coop.next_batch('testqueue', 'maincons', 'subcons2', '1 hour');

select pgq_coop.unregister_subconsumer('testqueue', 'maincons', 'subcons1', 0);
select pgq_coop.unregister_subconsumer('testqueue', 'maincons', 'subcons2', 0);
select pgq_coop.unregister_subconsumer('testqueue', 'maincons', 'subcons2', 1);
select pgq_coop.unregister_subconsumer('testqueue', 'maincons', 'subcons2', 0);

-- test auto-creation
select pgq_coop.next_batch('testqueue', 'cmain', 'sc1');
select pgq_coop.next_batch('testqueue', 'cmain', 'sc2');
select consumer_name, last_tick from pgq.get_consumer_info();

-- test unregistering with pure pgq api
select pgq.unregister_consumer('testqueue', 'cmain.sc2');
select pgq.unregister_consumer('testqueue', 'cmain');
select consumer_name, last_tick from pgq.get_consumer_info();

