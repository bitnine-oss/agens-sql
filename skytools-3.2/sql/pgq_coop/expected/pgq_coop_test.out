select pgq.create_queue('testqueue');
 create_queue 
--------------
            1
(1 row)

update pgq.queue set queue_ticker_max_count = 1 where queue_name = 'testqueue';
-- register
select pgq_coop.register_subconsumer('testqueue', 'maincons', 'subcons1');
 register_subconsumer 
----------------------
                    1
(1 row)

select pgq_coop.register_subconsumer('testqueue', 'maincons', 'subcons1');
 register_subconsumer 
----------------------
                    0
(1 row)

select pgq_coop.register_subconsumer('testqueue', 'maincons', 'subcons1');
 register_subconsumer 
----------------------
                    0
(1 row)

select pgq_coop.register_subconsumer('testqueue', 'maincons', 'subcons2');
 register_subconsumer 
----------------------
                    1
(1 row)

-- process events
select pgq_coop.next_batch('testqueue', 'maincons', 'subcons1');
 next_batch 
------------
           
(1 row)

select pgq.insert_event('testqueue', 'ev0', 'data');
 insert_event 
--------------
            1
(1 row)

select pgq.insert_event('testqueue', 'ev1', 'data');
 insert_event 
--------------
            2
(1 row)

select pgq.insert_event('testqueue', 'ev2', 'data');
 insert_event 
--------------
            3
(1 row)

select pgq.ticker();
 ticker 
--------
      1
(1 row)

select pgq_coop.next_batch('testqueue', 'maincons', 'subcons1');
 next_batch 
------------
          1
(1 row)

select pgq_coop.next_batch('testqueue', 'maincons', 'subcons1');
 next_batch 
------------
          1
(1 row)

select pgq_coop.next_batch('testqueue', 'maincons', 'subcons2');
 next_batch 
------------
           
(1 row)

select pgq.insert_event('testqueue', 'ev3', 'data');
 insert_event 
--------------
            4
(1 row)

select pgq.insert_event('testqueue', 'ev4', 'data');
 insert_event 
--------------
            5
(1 row)

select pgq.insert_event('testqueue', 'ev5', 'data');
 insert_event 
--------------
            6
(1 row)

select pgq.ticker();
 ticker 
--------
      1
(1 row)

select pgq_coop.next_batch('testqueue', 'maincons', 'subcons2');
 next_batch 
------------
          2
(1 row)

select pgq_coop.finish_batch(2);
 finish_batch 
--------------
            1
(1 row)

-- test takeover
select pgq_coop.next_batch('testqueue', 'maincons', 'subcons2', '1 hour');
 next_batch 
------------
           
(1 row)

update pgq.subscription set sub_active = '2005-01-01' where sub_batch is not null;
select pgq_coop.next_batch('testqueue', 'maincons', 'subcons2', '1 hour');
 next_batch 
------------
          1
(1 row)

select pgq_coop.unregister_subconsumer('testqueue', 'maincons', 'subcons1', 0);
 unregister_subconsumer 
------------------------
                      0
(1 row)

select pgq_coop.unregister_subconsumer('testqueue', 'maincons', 'subcons2', 0);
ERROR:  subconsumer has active batch
select pgq_coop.unregister_subconsumer('testqueue', 'maincons', 'subcons2', 1);
 unregister_subconsumer 
------------------------
                      1
(1 row)

select pgq_coop.unregister_subconsumer('testqueue', 'maincons', 'subcons2', 0);
 unregister_subconsumer 
------------------------
                      0
(1 row)

-- test auto-creation
select pgq_coop.next_batch('testqueue', 'cmain', 'sc1');
 next_batch 
------------
           
(1 row)

select pgq_coop.next_batch('testqueue', 'cmain', 'sc2');
 next_batch 
------------
           
(1 row)

select consumer_name, last_tick from pgq.get_consumer_info();
 consumer_name | last_tick 
---------------+-----------
 cmain         |         3
 cmain.sc1     |          
 cmain.sc2     |          
 maincons      |         3
(4 rows)

-- test unregistering with pure pgq api
select pgq.unregister_consumer('testqueue', 'cmain.sc2');
 unregister_consumer 
---------------------
                   1
(1 row)

select pgq.unregister_consumer('testqueue', 'cmain');
 unregister_consumer 
---------------------
                   2
(1 row)

select consumer_name, last_tick from pgq.get_consumer_info();
 consumer_name | last_tick 
---------------+-----------
 maincons      |         3
(1 row)

